# Pre-Lecture HW

### Question 1: 

In [1]:
import pandas as pd
url = "https://raw.githubusercontent.com/rfordatascience/tidytuesday/master/data/2020/2020-05-05/villagers.csv"
villagers = pd.read_csv(url)
villagers.isna().sum()

row_n           0
id              1
name            0
gender          0
species         0
birthday        0
personality     0
song           11
phrase          0
full_id         0
url             0
dtype: int64

In [2]:
url2 = 'https://raw.githubusercontent.com/mwaskom/seaborn-data/master/titanic.csv'
titanic = pd.read_csv(url2)
titanic.isna().sum()

survived         0
pclass           0
sex              0
age            177
sibsp            0
parch            0
fare             0
embarked         2
class            0
who              0
adult_male       0
deck           688
embark_town      2
alive            0
alone            0
dtype: int64

### Question 2: 

In [3]:
print(titanic.shape)
titanic

(891, 15)


,survived,pclass,sex,age,sibsp,parch,fare,embarked,class,who,adult_male,deck,embark_town,alive,alone
0,0,3,male,22.0,1,0,7.2500,S,Third,man,True,NaN,Southampton,no,False
1,1,1,female,38.0,1,0,71.2833,C,First,woman,False,C,Cherbourg,yes,False
2,1,3,female,26.0,0,0,7.9250,S,Third,woman,False,NaN,Southampton,yes,True
3,1,1,female,35.0,1,0,53.1000,S,First,woman,False,C,Southampton,yes,False
4,0,3,male,35.0,0,0,8.0500,S,Third,man,True,NaN,Southampton,no,True
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
886,0,2,male,27.0,0,0,13.0000,S,Second,man,True,NaN,Southampton,no,True
887,1,1,female,19.0,0,0,30.0000,S,First,woman,False,B,Southampton,yes,True
888,0,3,female,NaN,1,2,23.4500,S,Third,woman,False,NaN,Southampton,no,False
889,1,1,male,26.0,0,0,30.0000,C,First,man,True,C,Cherbourg,yes,True


In [4]:
print(villagers.shape)
villagers

(391, 11)


,row_n,id,name,gender,species,birthday,personality,song,phrase,full_id,url
0,2,admiral,Admiral,male,bird,1-27,cranky,Steep Hill,aye aye,villager-admiral,https://villagerdb.com/images/villagers/thumb/...
1,3,agent-s,Agent S,female,squirrel,7-2,peppy,DJ K.K.,sidekick,villager-agent-s,https://villagerdb.com/images/villagers/thumb/...
2,4,agnes,Agnes,female,pig,4-21,uchi,K.K. House,snuffle,villager-agnes,https://villagerdb.com/images/villagers/thumb/...
3,6,al,Al,male,gorilla,10-18,lazy,Steep Hill,Ayyeeee,villager-al,https://villagerdb.com/images/villagers/thumb/...
4,7,alfonso,Alfonso,male,alligator,6-9,lazy,Forest Life,it'sa me,villager-alfonso,https://villagerdb.com/images/villagers/thumb/...
...,...,...,...,...,...,...,...,...,...,...,...
386,475,winnie,Winnie,female,horse,1-31,peppy,My Place,hay-OK,villager-winnie,https://villagerdb.com/images/villagers/thumb/...
387,477,wolfgang,Wolfgang,male,wolf,11-25,cranky,K.K. Song,snarrrl,villager-wolfgang,https://villagerdb.com/images/villagers/thumb/...
388,480,yuka,Yuka,female,koala,7-20,snooty,Soulful K.K.,tsk tsk,villager-yuka,https://villagerdb.com/images/villagers/thumb/...
389,481,zell,Zell,male,deer,6-7,smug,K.K. D&B,pronk,villager-zell,https://villagerdb.com/images/villagers/thumb/...


Observation represents rows within a dataset. Generally, observation focuses on one entity of a dataset and all the information about that entity. In the animal crossing dataset though, one observation would be the entity "Agent S", and that represents all the information about "Agent S". Its birthday, species, phrase, etc... 

Variable represents columns of the dataset, where each column would store the same type of information, but information regarding different entities. For example, "id" is a variable that stores identification of every villager. It's the same type of data, but data from different observations. 

### Question 3 

In [6]:
titanic.describe(include='all')

,survived,pclass,sex,age,sibsp,parch,fare,embarked,class,who,adult_male,deck,embark_town,alive,alone
count,891.000000,891.000000,891,714.000000,891.000000,891.000000,891.000000,889,891,891,891,203,889,891,891
unique,NaN,NaN,2,NaN,NaN,NaN,NaN,3,3,3,2,7,3,2,2
top,NaN,NaN,male,NaN,NaN,NaN,NaN,S,Third,man,True,C,Southampton,no,True
freq,NaN,NaN,577,NaN,NaN,NaN,NaN,644,491,537,537,59,644,549,537
mean,0.383838,2.308642,NaN,29.699118,0.523008,0.381594,32.204208,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
std,0.486592,0.836071,NaN,14.526497,1.102743,0.806057,49.693429,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
min,0.000000,1.000000,NaN,0.420000,0.000000,0.000000,0.000000,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
25%,0.000000,2.000000,NaN,20.125000,0.000000,0.000000,7.910400,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
50%,0.000000,3.000000,NaN,28.000000,0.000000,0.000000,14.454200,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
75%,1.000000,3.000000,NaN,38.000000,1.000000,0.000000,31.000000,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN


In [58]:
titanic['deck'].value_counts()

deck
C    59
B    47
D    33
E    32
A    15
F    13
G     4
Name: count, dtype: int64

### Question 4

The total count for the 'age' variable is 714 and there are 891 rows for the data frame. The discrepency is probably due to data being missing or the data was never recorded. 
For the 'deck' varible, 688 pieces of data are missing, and this is probably due to the reason that only the survivors were able to give information on which deck they were on at the time of the collision, and many of the survivors probably don't remember either. Only the observations with 'survived' as '1', has data in their 'deck' variable (working with the few rows that are shown in jupyter). 

### Question 5

Attribute is something inherent to the object, it's an characteristic. It stores data for the object and parentheses aren't needed because it isn't a function call, it is just retrieving a value. 

Method is a type of function, more specifically it's a function that only works with that particular object. Method evaluates / performs actions on arguments given to it (not all methods or functions require arguments though). 

# Pre-Lecture HW Chat Summary 

We began by discussing a dataset of villagers, specifically from the URL you provided. After an initial issue accessing the dataset, we explored key concepts such as observations (rows representing individual villagers) and variables (columns representing attributes like species or personality).

You asked for code to summarize the dataset, and I explained how .describe() provides summaries for both numerical and categorical columns. When you encountered an issue with df.describe(include=number), I clarified that number should be replaced with either np.number (if using NumPy) or ['float64', 'int64'] for numeric columns using only Pandas.

We also compared .describe() with .value_counts(), noting that .describe() provides general statistics, while .value_counts() focuses on frequency counts of unique values in a single column.

Finally, we covered the difference between attributes (characteristics or properties of an object) and methods (functions that perform actions on an object).

# Post-Lecture HW

### Question 6

Count: Number of entries, not including 'NaN'
Mean: Average value of the entries 
Standard deviation (std): It measures how "spread out" or how much the data deviate from the mean
Minimum (min): Smallest value in a variable 
25th Percentile (25%): The value where 25% of the data falls, also called the first quartile
50th Percentile (50%): The value where half of the data falls, also known as the median
75th Percentile (75%): The value where 75% of the data fals, also called the third quartile
Maximum (max): Largest value in a variable

### Question 7

#### 1.

When someone would want to remove only the NaN value and nothing else. del df['col'] would delete the entire column. Also df.dropna() gives you more customizability, as you can also use it on rows, while del df['col'] can't be. 

#### 2.

When someone would just want to get rid of an entire column, and not leave data behind. 

#### 3. 

When someone would like to clense their data frame, it will be useful to have del df['col'] getting rid of all the irrelevant data, and then using df.dropna() to get rid of the rows/columns with NaN. 

#### 4.

In [7]:
titanic # before

,survived,pclass,sex,age,sibsp,parch,fare,embarked,class,who,adult_male,deck,embark_town,alive,alone
0,0,3,male,22.0,1,0,7.2500,S,Third,man,True,NaN,Southampton,no,False
1,1,1,female,38.0,1,0,71.2833,C,First,woman,False,C,Cherbourg,yes,False
2,1,3,female,26.0,0,0,7.9250,S,Third,woman,False,NaN,Southampton,yes,True
3,1,1,female,35.0,1,0,53.1000,S,First,woman,False,C,Southampton,yes,False
4,0,3,male,35.0,0,0,8.0500,S,Third,man,True,NaN,Southampton,no,True
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
886,0,2,male,27.0,0,0,13.0000,S,Second,man,True,NaN,Southampton,no,True
887,1,1,female,19.0,0,0,30.0000,S,First,woman,False,B,Southampton,yes,True
888,0,3,female,NaN,1,2,23.4500,S,Third,woman,False,NaN,Southampton,no,False
889,1,1,male,26.0,0,0,30.0000,C,First,man,True,C,Cherbourg,yes,True


In [ ]:
del titanic['deck'] #isn't retre

In [15]:
titanic.dropna(inplace=True)
titanic

,survived,pclass,sex,age,sibsp,parch,fare,embarked,class,who,adult_male,embark_town,alive,alone
0,0,3,male,22.0,1,0,7.2500,S,Third,man,True,Southampton,no,False
1,1,1,female,38.0,1,0,71.2833,C,First,woman,False,Cherbourg,yes,False
2,1,3,female,26.0,0,0,7.9250,S,Third,woman,False,Southampton,yes,True
3,1,1,female,35.0,1,0,53.1000,S,First,woman,False,Southampton,yes,False
4,0,3,male,35.0,0,0,8.0500,S,Third,man,True,Southampton,no,True
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
885,0,3,female,39.0,0,5,29.1250,Q,Third,woman,False,Queenstown,no,False
886,0,2,male,27.0,0,0,13.0000,S,Second,man,True,Southampton,no,True
887,1,1,female,19.0,0,0,30.0000,S,First,woman,False,Southampton,yes,True
889,1,1,male,26.0,0,0,30.0000,C,First,man,True,Cherbourg,yes,True


I decided to delete the 'deck' column right away because there are so many missing entries, so if I use titanic.dropna() the entire dataset might as well be gone. So after using del titanic['deck'] I used titanic.dropna(inplace=True) to modify the dataset directly instead of having a copy which I could have done. 

## Question 8

#### 1. 

In [9]:
titanic.groupby('alone')['survived'].describe()

KeyError: 'ALONE'

.groupby('col1')['col2'] functions by grouping all the unique entries of 'col1' together. And then it will only look at 'col2' and the data in 'col2' would be sorted based on the unique entries of 'col1'. Finally with .describe() it will generate all the statistical analysis values for the user to use. 

So in my example, I wanted to look at whether being alone on the titanic would impact your chances at survival. So I used .groupby('alone') to separate the loners from the groups of people. Then have .describe() analyze the ['survived'] in the loners and the groups. And now I can see the loners had a significantly worse chance of survival than the groups. 

#### 2.

Using .describe() initially is just listing out data for you. It doesn't mean anything, instead it's giving you facts that someone could use to derive a meaning. But with .groupby('col1')['col2'].describe() two different parts are being compared, that process of comparison leads to patterns and more information gained. Using my example above, I was able to find out that groups of people had a significantly higher chance of survival than people that were alone. Which, if we didn't use .groupby('col1')['col2'] would be impossible to see with just the fact that 38% of people survived from the collison. 

#### 3.

For some problems Chat and Google performed to the same level of proficiency, but for Google I have to actively search out answers, click on different links, and sometimes the link I fist click on doesn't resolve my issue so I have to do it again. For Chat it comes down to convenience, when I ask Chat something they answer me right away. And for these coding problems Chat is pretty good at detecting what's wrong. Maybe there is a bias due to the fact that I have been using the same conversation with Chat for my HW so Chat is in the STA130 frame of mind. Still it is way easier to copy and paste the problem into Chat than to go on Google to try to search for an answer. 

### Question 9

Somewhat. I didn't have any experience with pandas or jupyter or github so I was super confused. Especially after the first tutorial where nothing was really explained to me and was just thrown into the deepend. But Chat really helped me, any questions Chat would answer to a good degree of satisfaction. 

# Post-Lecture HW Chat Summary
